src/nsrt_learning/side_predicate_learning.py

```diff
@@ -394,59 +400,92 @@ def _sideline(self) -> List[PartialNSRTAndDatastore]:
                 segment = seg_traj[t]
                 option = segment.get_option()
                 # Find the PNAD associated with this option.
-                pnad = param_opt_to_pnad[option.parent]
+                if param_opt_to_nec_pnad.get(option.parent) is None:
+                    pnads_for_option = [
+                        param_opt_to_general_pnad[option.parent]
+                    ]
+                else:
+                    pnads_for_option = param_opt_to_nec_pnad[option.parent]
+
                 # Compute the ground atoms that must be added on this timestep.
                 # They must be a subset of the current PNAD's add effects.
                 necessary_add_effects = necessary_image - atoms_seq[t]
                 assert necessary_add_effects.issubset(segment.add_effects)
 
-                # Find a mapping from the variables in the PNAD add effects
-                # and option to the objects in necessary_add_effects and the
-                # segment's option. If one exists, we don't need to modify this
-                # PNAD. Otherwise, we must make its add effects more specific.
-                # Note that we are assuming all variables in the parameters
-                # of the PNAD will appear in either the option arguments or
-                # the add effects. This is in contrast to strips_learning.py,
-                # where delete effect variables also contribute to parameters.
-                opt_pred = Predicate("OPT-ARGS",
-                                     [a.type for a in pnad.option_spec[1]],
-                                     _classifier=lambda s, o: False)  # dummy
-                opt_vars_atom = frozenset(
-                    {LiftedAtom(opt_pred, pnad.option_spec[1])})
-                opt_objs_atom = frozenset(
-                    {GroundAtom(opt_pred, option.objects)})
-                unifies, obj_to_var_ent = utils.find_substitution(
-                    pnad.op.add_effects | opt_vars_atom,
-                    necessary_add_effects | opt_objs_atom)
-                obj_to_var = cast(ObjToVarSub, obj_to_var_ent)
-
-                if not unifies:  # we need to make add effects more specific
-                    del obj_to_var  # undefined; delete it for safety
-                    # We now try to make the PNAD's add effects more specific.
-                    # Since we have started with the most general operators,
-                    # and we are making them more specific, it is guaranteed
-                    # that such a grounding exists, otherwise our 1:1
-                    # option:operator assumption is violated.
-                    new_pnad = self._try_specializing_pnad(
-                        necessary_add_effects, pnad, segment)
-                    assert new_pnad is not None, "1:1 assumption violated"
-                    assert new_pnad.option_spec == pnad.option_spec
+                substitution_exists = False
```

Comment:
You don't need this `substitution_exists` variable; you can use the `for ... else ...` syntax instead (https://book.pythontips.com/en/latest/for_-_else.html) just change the `if not substitution_exists:` below to `else:`